quinn-proto/src/packet.rs

```diff
@@ -85,34 +111,71 @@ impl PartialDecode {
             }
             InvariantHeader::Long {
                 first,
-                version,
+                version: 0,
                 dst_cid,
                 src_cid,
-            } => {
-                if version == 0 {
+                ..
+            } => (
+                buf.remaining(),
+                Header::VersionNegotiate {
+                    random: first & !LONG_HEADER_FORM,
+                    src_cid,
+                    dst_cid,
+                },
+                false,
+            ),
+            InvariantHeader::Long {
+                first,
+                version: VERSION,
+                dst_cid,
+                src_cid,
+            } => match PacketType::from_byte(first)? {
+                PacketType::Retry => {
+                    let odcil = buf.get::<u8>()? as usize;
+                    let mut odci_stage = [0; 18];
+                    buf.copy_to_slice(&mut odci_stage[0..odcil]);
+                    let orig_dst_cid = ConnectionId::new(&odci_stage[..odcil]);
                     (
-                        buf.position() as usize,
                         buf.remaining(),
-                        Header::VersionNegotiate {
-                            random: first & !LONG_HEADER_FORM,
+                        Header::Retry {
                             src_cid,
                             dst_cid,
+                            orig_dst_cid,
                         },
                         false,
                     )
-                } else {
-                    debug_assert_eq!(version, VERSION);
-                    let ty = LongType::from_byte(first)?;
+                }
+                PacketType::Initial => {
+                    let token_length = buf.get_var()? as usize;
+                    let mut token = vec![0; token_length];
+                    buf.copy_to_slice(&mut token);
+
                     let len = buf.get_var()?;
-                    let number = buf.get()?;
-                    let header_len = buf.position() as usize;
-                    if len > buf.remaining() as u64 {
-                        return Err(PacketDecodeError::InvalidHeader(
-                            "payload longer than packet",
-                        ));
-                    }
+                    let sample_offset = 10
+                        + dst_cid.len()
+                        + src_cid.len()
+                        + varint::size(len).unwrap()
+                        + varint::size(token_length as u64).unwrap()
+                        + token.len();
```

Comment:
Probable missing bounds check. Perhaps the check/adjustment should be done inside `get_packet_number`?